hashes/zkevm/src/keccak/computation_circuit/circuit.rs

```diff
@@ -0,0 +1,343 @@
+    //     {bytes_left: N/A, word_value:                N/A, is_final:  true},
+    // ]
+    // For easier Poseidon processing, the raw input is encoded in the following way:
+    // If the logic input is empty, [0; RATE].
+    // If the logic input is empty, each word is encoded as RATE witnesses. First word is encoded
+    // as [<byte length of logical input>, logical_input[0], 0..]. Other words are encoded as [logical_input[i], 0, 0..].
+    // Note: With this encoding, App circuits should be able to compute Poseidons of variable length inputs easily.
+    // Then we get the following compact inputs for Poseidon hasher:
+    // poseidon_compact_inputs = [
+    //  {inputs: [                21, 0x0807060504030201], len: 2, is_final: false},   // 21 is the length of the logical input.
+    //  {inputs: [0x1817161514131211,                0x0], len: 2, is_final: false},
+    //  {inputs: [0x0000002524232221,                0x0], len: 2, is_final: true},    // The last row of the logical input.
+    //  {inputs: [               0x0,                0x0], len: 2, is_final: true},    // This row corresponds to the padding row in loaded_assigned_rows.
+    // ]
+    // The Poseidon results will be:
+    // poseidon_compact_outputs = [
+    //  {hash:                                N/A, is_final: false},
+    //  {hash:                                N/A, is_final: false},
+    //  {hash: <poseidon(<encoded logical input>), is_final:  true},
+    //  {hash:               poseidon([0x0, 0x0]), is_final:  true},
+    // ]
+    // Because of the padding rows, is_final cannot tell which row is the result of a logical input. Therefore
+    // we also build a selector array(2d, <num of chunks> * NUM_WORDS_TO_ABSORB) to select reuslts:
+    // poseidon_selector: [[0, 0], [1, 0]]
+    fn generate_poseidon_inputs(
+        &self,
+        loaded_assigned_rows: &[LoadedKeccakAssignedRow<F>],
+    ) -> (
+        Vec<PoseidonCompactInput<F, { poseidon_params::RATE }>>,
+        Vec<[AssignedValue<F>; NUM_WORDS_TO_ABSORB]>,
+    ) {
+        let rows_per_round = self.config_params.keccak_circuit_params.rows_per_round;
+        let mut range_circuit_builder_mut = self.range_circuit_builder.borrow_mut();
+        let ctx = range_circuit_builder_mut.main(0);
+        let range_chip = self.range_circuit_builder.borrow().range_chip();
+
+        let num_chunks = (loaded_assigned_rows.len() / rows_per_round - 1) / (NUM_ROUNDS + 1);
+        let compact_input_len = num_chunks * (NUM_WORDS_TO_ABSORB);
+        let mut result_selector_per_chunk = Vec::with_capacity(num_chunks);
+        let mut compact_inputs = Vec::with_capacity(compact_input_len);
+
+        let rate_witness = ctx.load_constant(F::from(poseidon_params::RATE as u64));
+        let zero_witness = ctx.load_zero();
+        let mut chunk_is_final_last = zero_witness;
+
+        // Skip the first round which is dummy.
+        for chunk in
```

Comment:
as mentioned above, all chunk processing and step_by should be done before loading external cells